src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -91,55 +83,75 @@ pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context
-async fn get_bit_equality_checkers<F, C, S>(
-    idx: usize,
-    multi_bit_input: &[Vec<S>],
+async fn check_everything<F, C, S>(
     ctx: C,
+    record_idx: usize,
+    input: &[Vec<S>],
 ) -> Result<Vec<S>, Error>
 where
     F: Field,
     C: Context<F, Share = S>,
     S: SecretSharing<F>,
 {
-    let num_multi_bits = multi_bit_input.len();
-    assert_ne!(num_multi_bits, 0);
-    let num_records = multi_bit_input[0].len();
-
-    let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);
-    let idx_in_bits = &idx.view_bits::<Lsb0>()[..num_multi_bits];
-
-    try_join_all(
-        (0..num_records)
-            .zip(repeat(ctx_across_bits))
-            .map(|(rec, ctx)| async move {
-                let share_of_one = ctx.share_of_one();
-
-                let mult_input = zip(multi_bit_input, idx_in_bits)
-                    .map(|(single_bit_input, bit)| {
-                        if *bit {
-                            single_bit_input[rec].clone()
-                        } else {
-                            -single_bit_input[rec].clone() + &share_of_one
-                        }
-                    })
-                    .collect::<Vec<_>>();
-
-                multiply_all_shares(
-                    ctx,
-                    RecordId::from(idx * num_records + rec),
-                    mult_input.as_slice(),
-                )
-                .await
-            }),
-    )
-    .await
+    let record_id = RecordId::from(record_idx);
+    let num_bits = input.len();
+    let mut precomputed_combinations = Vec::with_capacity(1 << num_bits);
+    precomputed_combinations.push(ctx.share_of_one());
+    #[allow(clippy::needless_range_loop)]
+    for bit_idx in 0..num_bits {
+        let bit = &input[bit_idx][record_idx];
+        let step = 1 << bit_idx;
+        let num_children_to_add = precomputed_combinations.len();
+        precomputed_combinations.push(bit.clone());
+        if bit_idx == 0 {
+            continue;
+        }
+        let mut multiplication_futures = Vec::with_capacity(num_children_to_add - 1);
+        #[allow(clippy::needless_range_loop)]
+        for j in 1..num_children_to_add {
+            let child_idx = j + step;
+            multiplication_futures.push(ctx.narrow(&BitOpStep::from(child_idx)).multiply(
+                record_id,
+                &precomputed_combinations[j],
+                bit,
+            ));
+        }
+        let mut multiplication_results = try_join_all(multiplication_futures).await?;
+        precomputed_combinations.append(&mut multiplication_results);
+    }
+
+    let mut equality_checks = Vec::with_capacity(1 << num_bits);
+    for i in 0..(1 << num_bits) {
+        equality_checks.push(check_equality_to(
+            i,
+            num_bits,
+            precomputed_combinations.as_slice(),
+            0,
+            0,
+        ));
+    }
+    Ok(equality_checks)
+}
+
+fn check_equality_to<F: Field, S: SecretSharing<F>>(
```

Comment:
This is obviously correct (the tests pass), but the code is incomprehensible to me.